sdk/dotnet/cmd/pulumi-language-dotnet/main.go

```diff
@@ -805,3 +814,104 @@ func (host *dotnetLanguageHost) GetProgramDependencies(
 		Dependencies: packages,
 	}, nil
 }
+
+func (host *dotnetLanguageHost) GenerateProject(
+	ctx context.Context, req *pulumirpc.GenerateProjectRequest) (*pulumirpc.GenerateProjectResponse, error) {
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+	sink := diag.DefaultSink(os.Stderr, os.Stderr, diag.FormatOptions{
+		Color: cmdutil.GetGlobalColorization(),
+	})
+	pluginCtx, err := plugin.NewContext(sink, sink, nil, nil, cwd, nil, true, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer contract.IgnoreClose(pluginCtx)
+
+	loader := schema.NewPluginLoader(pluginCtx.Host)
+	parser := hclsyntax.NewParser()
+	for file, source := range req.Source {
+		if err := parser.ParseFile(strings.NewReader(source), file); err != nil {
+			return nil, err
+		}
+		if parser.Diagnostics.HasErrors() {
+			return nil, parser.Diagnostics
+		}
+	}
+	bindOpts := []pcl.BindOption{
+		pcl.SkipResourceTypechecking,
+		pcl.AllowMissingProperties,
+		pcl.AllowMissingVariables,
+	}
+	bindOpts = append(bindOpts, pcl.Loader(loader))
+	program, pdiags, err := pcl.BindProgram(parser.Files, bindOpts...)
+	if err != nil {
+		return nil, err
+	}
+	if pdiags.HasErrors() {
+		return nil, pdiags
+	}
+
+	var project workspace.Project
+	err = encoding.JSON.Unmarshal([]byte(req.Project), &project)
+	if err != nil {
+		return nil, err
+	}
+
+	err = dotnetgen.GenerateProject(req.Directory, project, program, req.SourceOnly)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pulumirpc.GenerateProjectResponse{}, nil
+}
+
+func (host *dotnetLanguageHost) GeneratePackage(
+	ctx context.Context, req *pulumirpc.GeneratePackageRequest) (*pulumirpc.GeneratePackageResponse, error) {
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+	sink := diag.DefaultSink(os.Stderr, os.Stderr, diag.FormatOptions{
+		Color: cmdutil.GetGlobalColorization(),
+	})
+	pluginCtx, err := plugin.NewContext(sink, sink, nil, nil, cwd, nil, true, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer contract.IgnoreClose(pluginCtx)
+
+	var spec schema.PackageSpec
+	err = json.Unmarshal([]byte(req.Schema), &spec)
```

Comment:
Use segmentio/encoding/json and `json.Parse()` with the zerocopy option here.

Reply:
Should the protobuf Schema field just be `bytes` as well then? Seems a bit silly to copy the whole string to a byte slice to then invoke zero copy parsing on it.

Reply:
Yeah it should be, good catch. We're doing many full copies of strings in the schema here.

There is no way to prevent gRPC from copying though. gRPC will **always** copy the bytes received multiple times. The first copy is the scatter-gather from packets to put them contiguously into memory, the second copy is to protect the gRPC netcode from having its backing slice be modified.

Could we make this protocol take a schema URI instead of **or as an alternative to** schema bytes? If the schema URI is `file://` that would allow it to mmap, and if `https://` in the future would allow us to host schemas separately from plugins, e.g.: on our registry via CDN? That aligns with my vision for the registry: users run `pulumi package publish` which uploads a schema to our service, then we trigger jobs along the lines of `pulumi package sdkgen $schemaUri` and then publish those SDKs.

URIs instead of bytes would be fantastic here and decouple the protocol from caring about how the schema is transmitted.

Reply:
> Could we make this protocol take a schema URI instead of or as an alternative to schema bytes?

Well we can just always do a URI, if we're in one of the odd cases of having in-memory bytes on the engine side we can just spill to temp.

Reply:
True, spilling to temp might actually solve the perf issue for repeatedly generating examples well enough for providers that we don't need to do anything else. Our incremental and zero-copy schema parsing is really not that bad perf wise.